src/indieweb_utils/indieauth/flask.py

```diff
@@ -47,34 +67,35 @@ def indieauth_callback_handler(
         r = requests.post(token_endpoint, data=data, headers=headers)
     except:
         message = "Your token endpoint server could not be accessed."
-        return message, None
+        raise AuthenticationError(message)
 
     if r.status_code != 200:
         message = "There was an error with your token endpoint server."
-        return message, None
-
-    # remove code verifier from session because the authentication flow has finished
+        raise AuthenticationError(message)
 
     if me is None:
         message = "An invalid me value was provided."
-
-        return message, None
+        raise AuthenticationError(message)
 
     if r.json().get("me").strip("/") != me.strip("/"):
         message = "Your domain is not allowed to access this website."
-
-        return message
+        raise AuthenticationError(message)
 
     granted_scopes = r.json().get("scope").split(" ")
 
     if r.json().get("scope") == "" or any(scope not in granted_scopes for scope in required_scopes):
         message = f"You need to grant {', '.join(required_scopes).strip(', ')} access to use this tool."
-        return message, None
+        raise AuthenticationError(message)
```

Comment:
We could probably split this into a separate function that could be re-used for validating the response.